src/coreclr/vm/codeman.cpp

```diff
@@ -1929,6 +1950,120 @@ void EEJitManager::SetCpuInfo()
     CPUCompileFlags.Set64BitInstructionSetVariants();
     CPUCompileFlags.EnsureValidInstructionSetSupport();
 
+    const uint32_t smallestVectorBitWidth = 128;
+    const uint32_t largestVectorBitWidth = 2048;
+
+    uint32_t maxVectorTBitWidth      = CLRConfig::GetConfigValue(CLRConfig::EXTERNAL_MaxVectorTBitWidth);
+    uint32_t preferredVectorBitWidth = CLRConfig::GetConfigValue(CLRConfig::EXTERNAL_PreferredVectorBitWidth);
+
+#if defined(TARGET_X86) || defined(TARGET_AMD64)
+    CORINFO_InstructionSetFlags instructionSetFlags = CPUCompileFlags.GetInstructionSetFlags();
+
+    if (instructionSetFlags.HasInstructionSet(InstructionSet_AVX512BW_VL) &&
+        instructionSetFlags.HasInstructionSet(InstructionSet_AVX512CD_VL) &&
+        instructionSetFlags.HasInstructionSet(InstructionSet_AVX512DQ_VL))
+    {
+        assert(instructionSetFlags.HasInstructionSet(InstructionSet_AVX512BW));
+        assert(instructionSetFlags.HasInstructionSet(InstructionSet_AVX512CD));
+        assert(instructionSetFlags.HasInstructionSet(InstructionSet_AVX512DQ));
+        assert(instructionSetFlags.HasInstructionSet(InstructionSet_AVX512F));
+        assert(instructionSetFlags.HasInstructionSet(InstructionSet_AVX512F_VL));
+
+        if (isGenuineIntel)
+        {
+            // Some architectures can experience frequency throttling when executing
+            // executing 512-bit width instructions. To account for this we set the
+            // default preferred vector width to 256-bits in some scenarios. Users
+            // can override this with `DOTNET_PreferredVectorWith=512`.
```

Comment:
LLVM/GCC actually extend this `prefer-vector-width=256` behavior all the way up to the latest generation.

However, the general throttling issue has been fixed since Ice Lake and there are only a few instructions with "false dependencies" that can still cause some slowdown if used incorrectly. Since we don't have a general purpose auto-vectorizer and really just use this to control simple memory operations and comparisons, we should be fine limiting it to just the below.